File: packages/domain/src/discovery/signals.ts
/**
 * Strong / medium / weak signal extraction integrating query-pack classification (BB-039).
 */
import type { AdapterCandidateRecord } from '../adapters/types.js';
import {
  classifySignalStrength,
  evaluateTextAgainstTerms,
  type ClassifySignalResult,
} from '../query-packs/index.js';
import type { QueryPack, QueryTerm } from '../query-packs/types.js';
import type { DiscoverySignal } from './types.js';

function collectSearchableText(record: AdapterCandidateRecord): string {
  const parts: string[] = [];
  if (record.title) {
    parts.push(record.title);
  }
  if (record.classification) {
    parts.push(record.classification);
  }
  if (record.payload) {
    for (const value of Object.values(record.payload)) {
      if (typeof value === 'string') {
        parts.push(value);
      }
    }
  }
  return parts.join(' ');
}

export function matchTermsInCandidate(
  record: AdapterCandidateRecord,
  terms: readonly QueryTerm[],
): readonly QueryTerm[] {
  const text = collectSearchableText(record);
  return evaluateTextAgainstTerms(text, terms);
}

export function classifyDiscoverySignal(
  record: AdapterCandidateRecord,
  pack: QueryPack,
): ClassifySignalResult {
  const matchedTerms = matchTermsInCandidate(record, pack.terms);
  return classifySignalStrength({ matchedTerms });
}

export function toDiscoverySignal(
  result: ClassifySignalResult,
  matchedTerms: readonly QueryTerm[],
): DiscoverySignal {
  return {
    strength: result.strength,
    outcome: result.outcome,
    matchedClasses: result.matchedClasses,
    matchedTerms: matchedTerms.map((term) => term.text),
    reasons: result.reasons,
  };
}

export function extractDiscoverySignals(
  record: AdapterCandidateRecord,
  pack: QueryPack,
): DiscoverySignal {
  const matchedTerms = matchTermsInCandidate(record, pack.terms);
  return toDiscoverySignal(classifySignalStrength({ matchedTerms }), matchedTerms);
}

File: workers/research/src/black_book_research/discovery/signals.py
"""Signal extraction integrating query-pack classification (BB-039)."""

from __future__ import annotations

from black_book_research.adapters.types import AdapterCandidateRecord
from black_book_research.query_packs.classification import classify_signal_strength
from black_book_research.query_packs.pack import evaluate_text_against_terms
from black_book_research.query_packs.types import QueryPack

from .types import DiscoverySignal


def _collect_searchable_text(record: AdapterCandidateRecord) -> str:
    parts: list[str] = []
    if record.title:
        parts.append(record.title)
    if record.classification:
        parts.append(record.classification)
    if record.payload:
        for value in record.payload.values():
            if isinstance(value, str):
                parts.append(value)
    return " ".join(parts)


def extract_discovery_signals(record: AdapterCandidateRecord, pack: QueryPack) -> DiscoverySignal:
    text = _collect_searchable_text(record)
    matched_terms = evaluate_text_against_terms(text, pack.terms)
    result = classify_signal_strength(matched_terms=tuple(matched_terms))
    return DiscoverySignal(
        strength=result.strength,
        outcome=result.outcome,
        matched_classes=result.matched_classes,
        matched_terms=tuple(term.text for term in matched_terms),
        reasons=result.reasons,
    )
